src/ucp/proto/proto_am.inl

```diff
@@ -28,7 +28,9 @@ static UCS_F_ALWAYS_INLINE ucs_status_t ucp_am_handle_user_header_send_status(
 {
     ucs_status_t status;
 
-    if (ucs_unlikely(send_status == UCS_ERR_NO_RESOURCE) &&
+    if (ucs_likely(send_status == UCS_OK)) {
```

Comment:
But we need to check if header was copied before returning no? for example we tried to send it and received "UCS_ERR_NO_RESOURCE", than try to send it again and received UCS_OK. After the first attempt we might copied the header.

This was the logic behind this branch.
If "UCS_ERR_NO_RESOURCE" copy if needed
Any other status means that the request was completed or failed -> release header and return status.

Reply:
AFAIU if msg was sent (including the header), then no need to copy it at all.

Reply:
I think the question is if it was sent on the first attempt or in later attempts...

Reply:
I see your point. Anyway, once request completed and freed we can't access it, so `UCP_REQUEST_FLAG_USER_HEADER_COPIED` must be checked before this place, I'll see how we can refactor it...

Reply:
Not sure I understand. 
Calling "ucp_am_handle_user_header_send_status" happens before calling "ucp_request_complete" or i'm missing something?

Reply:
I see... that is a problem...
Need to fix it.
Is this the only place "ucp_am_handle_user_header_send_status" is called after "ucp_request_complete"?

Reply:
> Is this the only place "ucp_am_handle_user_header_send_status" is called after "ucp_request_complete"?

no

Reply:
ucp_am_eager_single_bcopy_reply_proto_progress?

If this are the only places (I didn't find more) maybe we need to call the header handler in "ucp_proto_am_bcopy_single_progress" after the send?

Reply:
Before this PR, send_status==UCS_OK would go to checking "req->flags & UCP_REQUEST_FLAG_USER_HEADER_COPIED" and release the mpool buffer if this check is true.
After this PR, send_status==UCS_OK would just return UCS_OK.
So where would we release the mpool buffer after this PR, if the header was copied to it?

Reply:
yes, this is what we discussed with @ofirfarjun7 above, I'm looking how better to fix it.

Reply:
I think we need to release `user_header` on completion. I found another path where we lose it, this is `abort`. So, it could be common place to do that in most cases @ofirfarjun7 @yosefe WDYT?

Reply:
you mean release it in ucp_request_complete?

Reply:
> you mean release it in ucp_request_complete?

no, it depends per protocol